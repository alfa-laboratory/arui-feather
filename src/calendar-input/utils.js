import getTime from 'date-fns/get_time';
import startOfDay from 'date-fns/start_of_day';
import formatDate from 'date-fns/format';
import isDateValid from 'date-fns/is_valid';
import { parse } from '../lib/date-utils';

/**
 * Разбирает введенную пользователем дату используя заданный формат.
 *
 * @param {String} value Дата
 * @param {String} format Ожидаемый формат даты
 * @returns {Number}
 */
export function parseDate(value, format) {
    const valueTrimmed = value ? value.replace(/~+$/, '') : '';
    let result = null;

    // Проверяем, чтобы пользователь ввёл полную строку даты без пробелов.
    if (valueTrimmed.length === format.length && !valueTrimmed.match(/\s/)) {
        let valueDate = parse(valueTrimmed, format);
        if (isDateValid(valueDate)) {
            result = valueDate.valueOf();
        }
    }

    return result;
}

/**
 * Изменяет формат даты с одного на другой.
 *
 * @param {String} value Оригинальная строка с датой
 * @param {String} inFormat Входной формат даты
 * @param {String} outFormat Формат возвращаемой даты
 * @returns {String}
 */
export function changeDateFormat(value, inFormat, outFormat) {
    let date = parseDate(value, inFormat);

    if (date) {
        return formatDate(date, outFormat);
    }
    return value;
}

/**
 * Возвращает дату с корректным месяцем.
 *
 * @param {String} value Строка даты
 * @param {String} format Формат строки с датой
 * @param {Number} [earlierLimit] Левая граница дат
 * @param {Number} [laterLimit] Правая граница дат
 * @returns {Number}
 */
export function calculateMonth(value, format, earlierLimit, laterLimit) {
    let newValue = value;

    if (typeof newValue === 'string') {
        newValue = parseDate(newValue, format);
    }

    if (!newValue || newValue !== newValue) { // eslint-disable-line no-self-compare
        newValue = Date.now();
    } else {
        newValue = getTime(newValue);
    }

    if (earlierLimit && earlierLimit > newValue) {
        return startOfDay(earlierLimit).valueOf();
    }
    if (laterLimit && laterLimit < newValue) {
        return startOfDay(laterLimit).valueOf();
    }

    return startOfDay(newValue).valueOf();
}
