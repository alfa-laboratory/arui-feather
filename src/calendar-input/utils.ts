import getTime from 'date-fns/get_time';
import startOfDay from 'date-fns/start_of_day';
import formatDate from 'date-fns/format';
import isDateValid from 'date-fns/is_valid';
import { parse } from '../lib/date-utils';

/**
 * Разбирает введенную пользователем дату используя заданный формат.
 *
 * @param value Дата
 * @param format Ожидаемый формат даты
 */
export function parseDate(value: string, format: string): number | null {
    const valueTrimmed = value ? value.replace(/~+$/, '') : '';
    let result = null;

    // Проверяем, чтобы пользователь ввёл полную строку даты без пробелов.
    if (valueTrimmed.length === format.length && !valueTrimmed.match(/\s/)) {
        const valueDate = parse(valueTrimmed, format);

        if (isDateValid(valueDate)) {
            result = valueDate.valueOf();
        }
    }

    return result;
}

/**
 * Изменяет формат даты с одного на другой.
 *
 * @param value Оригинальная строка с датой
 * @param inFormat Входной формат даты
 * @param outFormat Формат возвращаемой даты
 */
export function changeDateFormat(value: string, inFormat: string, outFormat: string) {
    const date = parseDate(value, inFormat);

    if (date) {
        return formatDate(date, outFormat);
    }

    return value;
}

/**
 * Возвращает дату с корректным месяцем.
 *
 * @param value Строка даты
 * @param format Формат строки с датой
 * @param earlierLimit Левая граница дат
 * @param laterLimit Правая граница дат
 */
export function calculateMonth(value: string, format: string, earlierLimit?: number, laterLimit?: number) {
    let newValue: string | number = value;

    if (typeof newValue === 'string') {
        newValue = parseDate(newValue, format);
    }

    if (!newValue || newValue !== newValue) { // eslint-disable-line no-self-compare
        newValue = Date.now();
    } else {
        newValue = getTime(newValue);
    }

    if (earlierLimit && earlierLimit > newValue) {
        return startOfDay(earlierLimit).valueOf();
    }
    if (laterLimit && laterLimit < newValue) {
        return startOfDay(laterLimit).valueOf();
    }

    return startOfDay(newValue).valueOf();
}

/**
 * Возвращает `true`, если поддерживается `input[type="date"]`
 */
export function isInputDateSupported() {
    const input = document.createElement('input');
    const value = 'a';

    input.setAttribute('type', 'date');
    input.setAttribute('value', value);

    return input.value !== value;
}
