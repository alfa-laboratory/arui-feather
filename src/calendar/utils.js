import startOfToday from 'date-fns/start_of_today';
import getTime from 'date-fns/get_time';

/**
 * Возвращает `true`, если переданная дата является текущей датой.
 *
 * @param {Data|Number} date Дата для проверки
 * @returns {Boolean}
 */
export function isCurrentDay(date) {
    if (date !== null) {
        return date.valueOf() === getTime(startOfToday());
    }

    return false;
}

/**
 * Возвращает массив - перечень лет, попадающих в переданный диапазон дат.
 *
 * @param {Data|Number} earlierLimit Начало диапазона дат
 * @param {Data|Number} laterLimit Окончание диапазона дат
 * @returns {Array} Диапазон лет
 */
export function getYearsRange(earlierLimit, laterLimit) {
    const years = [];
    const startYear = new Date(earlierLimit).getFullYear();
    const endYear = new Date(laterLimit).getFullYear();

    for (let currentYear = startYear; currentYear <= endYear; currentYear += 1) {
        years.push(currentYear);
    }

    return years.reverse();
}
