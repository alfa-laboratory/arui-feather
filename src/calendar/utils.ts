import startOfToday from 'date-fns/start_of_today';
import getTime from 'date-fns/get_time';

/**
 * Возвращает `true`, если переданная дата является текущей датой.
 *
 * @param date Дата для проверки
 */
export function isCurrentDay(
    date: Date | number
): boolean {
    if (date !== null) {
        return date.valueOf() === getTime(startOfToday());
    }

    return false;
}

/**
 * Возвращает массив - перечень лет, попадающих в переданный диапазон дат.
 * @param earlierLimit Начало диапазона дат
 * @param laterLimit Окончание диапазона дат
 * @returns Диапазон лет
 */
export function getYearsRange(
    earlierLimit: Date | number,
    laterLimit: Date | number
) {
    const years = [];
    const startYear = new Date(earlierLimit).getFullYear();
    const endYear = new Date(laterLimit).getFullYear();

    for (let currentYear = startYear; currentYear <= endYear; currentYear += 1) {
        years.push(currentYear);
    }

    return years.reverse();
}
