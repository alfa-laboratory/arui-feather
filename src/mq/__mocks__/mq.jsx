/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at http://mozilla.org/MPL/2.0/. */

/* eslint-disable react/prop-types */
import React from 'react';

let isMatched = true;

class Mq extends React.Component {
    componentDidMount() {
        this.props.onMatchChange(isMatched);
    }

    render() {
        return isMatched ? this.props.children : false;
    }
}

export function setIsMatched(newIsMatched) {
    isMatched = newIsMatched;
}

export default Mq;
