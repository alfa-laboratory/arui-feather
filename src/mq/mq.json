{
    "--small": "screen",
    "--small-only": "screen and (max-width: 47.9375em)",
    "--medium": "screen and (min-width: 48em)",
    "--medium-only": "screen and (min-width: 48em) and (max-width: 63.9375em)",
    "--large": "screen and (min-width: 64em)",
    "--large-only": "screen and (min-width: 64em) and (max-width: 89.9375em)",
    "--xlarge": "screen and (min-width: 90em)",
    "--xlarge-only": "screen and (min-width: 90em) and (max-width: 119.9375em)",
    "--xxlarge": "screen and (min-width: 120em)",
    "--xxlarge-only": "screen and (min-width: 120em) and (max-width: 99999999em)",
    "--mobile-s": "screen and (min-width: 20em)",
    "--mobile-m": "screen and (min-width: 23.4375em)",
    "--mobile-l": "screen and (min-width: 25.75em)",
    "--mobile": "screen and (max-width: 37.4375em)",
    "--tablet-s": "screen and (min-width: 37.5em)",
    "--tablet-m": "screen and (min-width: 48em)",
    "--tablet": "screen and (min-width: 37.5em) and (max-width: 63.9375em)",
    "--desktop-s": "screen and (min-width: 64em)",
    "--desktop-m": "screen and (min-width: 80em)",
    "--desktop-l": "screen and (min-width: 90em)",
    "--desktop-xl": "screen and (min-width: 120em)",
    "--desktop": "screen and (min-width: 64em)"
}
