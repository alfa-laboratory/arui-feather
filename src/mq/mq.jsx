/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at http://mozilla.org/MPL/2.0/. */

import autobind from 'core-decorators/lib/autobind';
import React from 'react';
import Type from 'prop-types';
import Modernizr from '../modernizr';
import { getMatchMedia, releaseMatchMedia } from '../lib/match-media';

const IS_BROWSER = typeof window !== 'undefined';
const SUPPORTS_TOUCH = IS_BROWSER && (Modernizr.pointerevents || Modernizr.touchevents);

/**
 * Компонент, имплементирующий поддержку медиа запросов в шаблонах.
 * Рендерит внутренние компоненты/разметку исходя из соответствия условиям запроса.
 * Для `query` используется window.matchMedia с полифиллом для IE9.
 * Можно использовать кастомные запросы из `src/mq/mq.json`, например `--small`.
 * Пока браузеры не поддерживают CSS4 Media Queries, поддержка тач-событий определяется через `touch`.
 */
class Mq extends React.Component {
    static propTypes = {
        /** Медиа запрос */
        query: Type.string,
        /** Запрос на поддержку тач-событий */
        touch: Type.bool,
        /** Дочерние элементы `Mq` */
        children: Type.node,
        /** Обработчик изменений в совпадении запросов */
        onMatchChange: Type.func
    };

    state = {
        isMatched: false
    };

    mql = null;

    componentDidMount() {
        this.mql = getMatchMedia(this.props.query);
        this.mql.addListener(this.handleMatch);
        this.handleMatch(this.mql);
    }

    componentWillUnmount() {
        releaseMatchMedia(this.props.query);
        this.mql.removeListener(this.handleMatch);
        this.mql = null;
    }

    render() {
        if (!this.props.children || !IS_BROWSER || !this.state.isMatched) return false;

        return this.props.children;
    }

    /**
     * @param {Object} [mql] MediaQueryList или MediaQueryListEvent
     */
    @autobind
    handleMatch(mql) {
        let queryPass = true;
        let touchPass = true;

        if (this.props.query) {
            queryPass = mql.matches;
        }
        if (this.props.touch) {
            touchPass = SUPPORTS_TOUCH;
        } else if (this.props.touch === false) {
            touchPass = !SUPPORTS_TOUCH;
        }

        this.setState({
            isMatched: (queryPass && touchPass)
        }, () => {
            if (this.props.onMatchChange) {
                this.props.onMatchChange(this.state.isMatched);
            }
        });
    }
}

export default Mq;
