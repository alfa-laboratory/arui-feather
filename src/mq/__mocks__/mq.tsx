/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at http://mozilla.org/MPL/2.0/. */

import React from 'react';

import { MqProps } from '../mq';

type MqMockProps = {
    onMatchChange: MqProps['onMatchChange'];
}

let isMatched = true;

class Mq extends React.Component<MqMockProps> {
    componentDidMount() {
        this.props.onMatchChange(isMatched);
    }

    render() {
        return isMatched ? this.props.children : false;
    }
}

export function setIsMatched(newIsMatched: boolean) {
    isMatched = newIsMatched;
}

export default Mq;
