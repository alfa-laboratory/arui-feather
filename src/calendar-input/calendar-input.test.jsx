/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at http://mozilla.org/MPL/2.0/. */

/* eslint import/no-extraneous-dependencies: [2, {"devDependencies": true}] */

import React from 'react';
import { shallow, mount } from 'enzyme';
import timezoneMock from 'timezone-mock';

import CalendarInput from './calendar-input';
import * as calendarUtils from './utils';
import keyboardCode from '../lib/keyboard-code';
import { SCROLL_TO_CORRECTION } from '../vars';
import { setIsMatched as setMqMatched } from '../mq/mq';

jest.mock('../mq/mq');

describe('calendar-input', () => {
    let originalWindowScrollTo = window.scrollTo;

    beforeAll(() => {
        timezoneMock.register('UTC');
    });

    afterAll(() => {
        timezoneMock.unregister();
    });

    beforeEach(() => {
        window.scrollTo = jest.fn();
        setMqMatched(false);
    });

    afterEach(() => {
        window.scrollTo = originalWindowScrollTo;
    });

    it('should render without problems', () => {
        const month = new Date('2018-06-15').valueOf();
        let calendarInput = shallow(<CalendarInput value='01.06.2018' calendar={ { month } } />);

        expect(calendarInput).toMatchSnapshot();
    });

    it('should render with calendar icon by default', () => {
        let calendarInput = mount(<CalendarInput />);
        let iconNode = calendarInput.find('IconButton');

        expect(iconNode.length).toBe(1);
    });

    it('should render without calendar icon with withIcon=false', () => {
        let calendarInput = mount(<CalendarInput withIcon={ false } />);
        let iconNode = calendarInput.find('IconButton');

        expect(iconNode.length).toBe(0);
    });

    it('should call `onInputChange` callback after input value was changed', () => {
        let onInputChange = jest.fn();
        let calendarInput = mount(<CalendarInput onInputChange={ onInputChange } />);


        calendarInput.find('input').simulate('change');

        expect(onInputChange).toHaveBeenCalled();
    });

    it('should call `onChange` callback after input value was changed', () => {
        let onChange = jest.fn();
        let calendarInput = mount(<CalendarInput onChange={ onChange } />);

        calendarInput.find('input').simulate('change');

        expect(onChange).toHaveBeenCalled();
    });

    it('should focus input on after `focus` call', () => {
        let calendarInput = mount(<CalendarInput />);
        let focusTarget = calendarInput.instance().customCalendarTarget;
        jest.spyOn(focusTarget, 'focus');

        calendarInput.instance().focus();

        expect(focusTarget.focus).toHaveBeenCalled();
    });

    it('should blur input on after `blur` call', () => {
        let calendarInput = mount(<CalendarInput />);
        let focusTarget = calendarInput.instance().customCalendarTarget;
        jest.spyOn(focusTarget, 'blur');

        calendarInput.instance().blur();

        expect(focusTarget.blur).toHaveBeenCalled();
    });

    it('should scroll window to element on public scrollTo method', () => {
        let calendarInput = mount(<CalendarInput />);
        let elemTopPosition = calendarInput.getDOMNode().getBoundingClientRect().top;
        let elemScrollTo = (elemTopPosition + window.pageYOffset) - SCROLL_TO_CORRECTION;

        calendarInput.instance().scrollTo();

        expect(window.scrollTo).toHaveBeenCalledWith(0, elemScrollTo);
    });

    it('should receive SyntheticEvent with type blur from input in first argument of `onInputFocus` callback', () => {
        let onInputFocus = jest.fn();
        let calendarInput = mount(<CalendarInput onInputFocus={ onInputFocus } />);

        calendarInput.find('input').simulate('focus');

        expect(onInputFocus).toHaveBeenCalled();
        expect(onInputFocus).toHaveBeenCalledWith(expect.objectContaining({ type: 'focus' }));
    });

    it('should receive SyntheticEvent with type blur from input in first argument of `onInputBlur` callback', () => {
        let onInputBlur = jest.fn();
        let calendarInput = mount(<CalendarInput onInputBlur={ onInputBlur } />);

        calendarInput.find('input').simulate('blur');

        expect(onInputBlur).toHaveBeenCalled();
        expect(onInputBlur).toHaveBeenCalledWith(expect.objectContaining({ type: 'blur' }));
    });

    it('should receive SyntheticEvent with type focus from component in first argument of `onFocus` callback', () => {
        let onFocus = jest.fn();
        let calendarInput = mount(<CalendarInput onFocus={ onFocus } />);

        calendarInput.find('input').simulate('focus');

        expect(onFocus).toHaveBeenCalled();
        expect(onFocus).toHaveBeenCalledWith(expect.objectContaining({ type: 'focus' }));
    });

    it('should receive SyntheticEvent with type blur from component in first argument of `onBlur` callback', () => {
        let onBlur = jest.fn();
        let calendarInput = mount(<CalendarInput onBlur={ onBlur } />);

        calendarInput.find('input').simulate('blur');

        expect(onBlur).toHaveBeenCalled();
        expect(onBlur).toHaveBeenCalledWith(expect.objectContaining({ type: 'blur' }));
    });


    it('should receive custom formatted date from event.target.value on `onFocus` callback', () => {
        let onFocus = jest.fn();
        let calendarInput = mount(<CalendarInput onFocus={ onFocus } value='01.08.2016' />);

        calendarInput.find('input').simulate('focus');

        expect(onFocus).toHaveBeenCalled();
        expect(onFocus.mock.calls[0][0].target.value).toEqual('01.08.2016');
        // expect(onFocus).toHaveBeenCalledWith({ target: { value: '01.08.2016' } });
    });

    it('should receive custom formatted date from event.target.value on `onBlur` callback', () => {
        let onBlur = jest.fn();
        const wrapper = mount(<CalendarInput onBlur={ onBlur } value='01.08.2016' />);

        wrapper.find('input').simulate('blur');

        expect(onBlur.mock.calls[0][0].target.value).toEqual('01.08.2016');
        // expect(onBlur).toHaveBeenCalledWith(expect.objectContaining({ target: { value: '01.08.2016' } }));
    });

    it('should receive custom formatted date from event.target.value on `onChange` callback', () => {
        let onChange = jest.fn();
        const wrapper = mount(<CalendarInput onChange={ onChange } />);
        const inputNode = wrapper.find('input');

        inputNode.simulate('change', { target: { value: '01.08.2016' } });

        expect(onChange).toHaveBeenCalledWith('01.08.2016', expect.any(Number));
    });

    it('should receive custom formatted date from event.target.value on `onInputChange` callback', () => {
        let onInputChange = jest.fn();
        const wrapper = mount(<CalendarInput onInputChange={ onInputChange } />);
        const inputNode = wrapper.find('input');

        inputNode.simulate('change', { target: { value: '01.08.2016' } });

        expect(onInputChange).toHaveBeenCalledWith('01.08.2016');
    });

    it('should open calendar after input was focused', (done) => {
        const wrapper = mount(<CalendarInput />);
        const calendarInput = wrapper.find('input');

        calendarInput.simulate('focus');

        setTimeout(() => {
            wrapper.update();
            expect(wrapper.find('Popup').props().visible).toBe(true);
            done();
        }, 0);
    });

    it('should call `onCalendarChange` callback after calendar value was changed', () => {
        let onCalendarChange = jest.fn();
        const wrapper = mount(<CalendarInput opened={ true } onCalendarChange={ onCalendarChange } />);
        const dayNode = wrapper.find('.calendar__day').at(15);

        dayNode.simulate('click');

        expect(onCalendarChange).toHaveBeenCalled();
    });

    it('should call `onChange` callback after calendar value was changed', () => {
        let onChange = jest.fn();
        const wrapper = mount(<CalendarInput opened={ true } onChange={ onChange } />);
        const dayNode = wrapper.find('.calendar__day').at(15);

        dayNode.simulate('click');

        expect(onChange).toHaveBeenCalled();
    });

    it('should close calendar popup after calendar value was changed by mouse click', () => {
        const wrapper = mount(<CalendarInput />);

        wrapper.setState({ isOpen: true });

        const dayNode = wrapper.find('.calendar__day').at(15);

        dayNode.simulate('click');

        expect(wrapper.find('Popup').props().visible).not.toBe(true);
    });

    it('should open calendar popup after down key was pressed in input', () => {
        const wrapper = mount(<CalendarInput />);
        const inputNode = wrapper.find('input');

        inputNode.simulate('keyDown', { which: keyboardCode.DOWN_ARROW });

        expect(wrapper.find('Popup').props().visible).toBe(true);
    });

    it('should close calendar popup after escape key was pressed in input', () => {
        const wrapper = mount(<CalendarInput />);
        const inputNode = wrapper.find('input');

        wrapper.setState({ isOpen: true });
        inputNode.simulate('keyDown', { which: keyboardCode.ESCAPE });

        expect(wrapper.find('Popup').props().visible).toBe(false);
    });

    it('should close calendar popup after tab key was pressed in input', () => {
        const wrapper = mount(<CalendarInput />);

        wrapper.setState({ isOpen: true });
        const inputNode = wrapper.find('input');

        inputNode.simulate('keyDown', { which: keyboardCode.TAB });

        expect(wrapper.find('Popup').props().visible).toBe(false);
    });

    it('should close calendar popup after enter or space key was pressed in calendar', () => {
        const wrapper = mount(<CalendarInput />);

        const calendarNode = wrapper.find('.calendar');

        wrapper.setState({ isOpen: true });
        calendarNode.simulate('keyDown', { which: keyboardCode.ESCAPE });

        expect(wrapper.find('Popup').props().visible).toBe(false);

        wrapper.setState({ isOpen: true });
        calendarNode.simulate('keyDown', { which: keyboardCode.ESCAPE });

        expect(wrapper.find('Popup').props().visible).toBe(false);
    });

    it('should focus on input after escape key was pressed in calendar', () => {
        const wrapper = mount(<CalendarInput />);
        const calendarTarget = wrapper.instance().customCalendarTarget;
        const calendarNode = wrapper.find('.calendar');
        jest.spyOn(calendarTarget, 'focus');

        calendarNode.simulate('keyDown', { which: keyboardCode.ESCAPE });

        expect(calendarTarget.focus).toHaveBeenCalled();
    });

    it('should focus on input after calendar icon was clicked', () => {
        const wrapper = mount(<CalendarInput />);
        const calendarTarget = wrapper.instance().customCalendarTarget;
        jest.spyOn(calendarTarget, 'focus');
        let iconNode = wrapper.find('.icon');

        iconNode.simulate('click');

        expect(calendarTarget.focus).toHaveBeenCalled();
    });

    it('should call `onCalendarKeyDown` callback after any key was pressed in calendar', () => {
        let onCalendarKeyDown = jest.fn();
        const wrapper = mount(<CalendarInput onCalendarKeyDown={ onCalendarKeyDown } />);
        const calendarNode = wrapper.find('.calendar');

        calendarNode.simulate('keyDown', { which: keyboardCode.NUMBER_0 });

        expect(onCalendarKeyDown).toHaveBeenCalled();
    });

    it('should call `onInputKeyDown` callback after any key was pressed in input', () => {
        let onInputKeyDown = jest.fn();
        const wrapper = mount(<CalendarInput onInputKeyDown={ onInputKeyDown } />);
        const inputNode = wrapper.find('input');

        inputNode.simulate('keyDown', { which: keyboardCode.NUMBER_0 });

        expect(onInputKeyDown).toHaveBeenCalled();
    });

    it('should call `onKeyDown` callback after any key was pressed in input and in calendar', () => {
        let onKeyDown = jest.fn();
        const wrapper = mount(<CalendarInput onKeyDown={ onKeyDown } />);
        const inputNode = wrapper.find('input');
        const calendarNode = wrapper.find('.calendar');

        inputNode.simulate('keyDown', { which: keyboardCode.NUMBER_0 });
        calendarNode.simulate('keyDown', { which: keyboardCode.NUMBER_0 });

        expect(onKeyDown).toHaveBeenCalledTimes(2);
    });

    describe('mobile', () => {
        beforeEach(() => {
            setMqMatched(true);
        });

        it('should set `max` attribute to `date`', () => {
            let calendar = {
                laterLimit: 1514505600000
            };

            const wrapper = mount(<CalendarInput calendar={ calendar } />);
            const inputNode = wrapper.find('input.calendar-input__native-control');

            expect(inputNode.props().max).toBe('2017-12-29');
        });

        it('should set `min` attribute to `date`', () => {
            let calendar = {
                earlierLimit: 1513900800000
            };

            const wrapper = mount(<CalendarInput calendar={ calendar } />);
            const inputNode = wrapper.find('input.calendar-input__native-control');

            expect(inputNode.props().min).toBe('2017-12-22');
        });
    });

    describe('calendar utils', () => {
        it('should change format of a date', () => {
            const result = calendarUtils.changeDateFormat('2012-11-10', 'YYYY-MM-DD', 'DD.MM.YYYY');
            expect(result).toBe('10.11.2012');
        });

        it('should return start of month', () => {
            const result = new Date(calendarUtils.calculateMonth('2012-11-10', 'YYYY-MM-DD'));
            expect(result.getMonth() + 1).toBe(11); // getMonth is zero based
            expect(result.getFullYear()).toBe(2012);
        });

        it('should return current month if not valid value given', () => {
            const result = new Date(calendarUtils.calculateMonth('foo', 'YYYY-MM-DD'));
            const now = new Date();
            expect(result.getMonth()).toBe(now.getMonth());
            expect(result.getFullYear()).toBe(now.getFullYear());
        });

        it('should return earlierLimit month if it after given date', () => {
            const result = new Date(calendarUtils.calculateMonth(
                '2012-11-10',
                'YYYY-MM-DD',
                (new Date(2013, 8, 10).getTime())
            ));
            expect(result.getMonth()).toBe(8);
            expect(result.getFullYear()).toBe(2013);
        });

        it('should return laterLimit month if it before given date', () => {
            const result = new Date(calendarUtils.calculateMonth(
                '2012-11-10',
                'YYYY-MM-DD',
                (new Date(2011, 8, 10).getTime()),
                (new Date(2011, 9, 10).getTime())
            ));
            expect(result.getMonth()).toBe(9);
            expect(result.getFullYear()).toBe(2011);
        });

        it('should return start of month if earlier and later limit given, but value is between them', () => {
            const result = new Date(calendarUtils.calculateMonth(
                '2012-11-10',
                'YYYY-MM-DD',
                (new Date(2011, 8, 10).getTime()),
                (new Date(2014, 9, 10).getTime())
            ));
            expect(result.getMonth() + 1).toBe(11); // getMonth is zero based
            expect(result.getFullYear()).toBe(2012);
        });
    });
});
