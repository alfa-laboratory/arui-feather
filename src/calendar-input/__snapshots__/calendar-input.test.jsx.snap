// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`calendar-input should render without problems 1`] = `
<span
  className="calendar-input calendar-input_theme_alfa-on-white"
>
  <span>
    <Mq
      onMatchChange={[Function]}
      query="--small-only"
      touch={true}
    >
      <input
        className="calendar-input__native-control"
        formNoValidate={true}
        onBlur={[Function]}
        onChange={[Function]}
        onFocus={[Function]}
        type="date"
        value="2018-06-01"
      />
    </Mq>
    <Input
      className="calendar-input__custom-control"
      disabledAttr={false}
      focused={false}
      formNoValidate={true}
      icon={
        <IconButton
          formNoValidate={false}
          onClick={[Function]}
          size="m"
          tag="button"
          type="button"
        >
          <IconCalendar />
        </IconButton>
      }
      mask="11.11.1111"
      onBlur={[Function]}
      onChange={[Function]}
      onFocus={[Function]}
      onKeyDown={[Function]}
      pattern="[0-9]*"
      placeholder="00.00.0000"
      size="m"
      type="tel"
      value="01.06.2018"
    />
  </span>
  <Popup
    directions={
      Array [
        "bottom-left",
        "bottom-right",
        "top-left",
        "top-right",
      ]
    }
    fitContaiterOffset={0}
    header={false}
    padded={true}
    secondaryOffset={0}
    size="s"
    target="anchor"
    visible={false}
  >
    <div
      className="calendar-input__calendar-wrapper"
    >
      <Calendar
        eventDays={Array []}
        isKeyboard={true}
        month={1529020800000}
        months={
          Array [
            "Январь",
            "Февраль",
            "Март",
            "Апрель",
            "Май",
            "Июнь",
            "Июль",
            "Август",
            "Сентябрь",
            "Октябрь",
            "Ноябрь",
            "Декабрь",
          ]
        }
        offDays={Array []}
        onBlur={[Function]}
        onFocus={[Function]}
        onKeyDown={[Function]}
        onMonthChange={[Function]}
        onValueChange={[Function]}
        outputFormat="DD.MM.YYYY"
        selectedFrom={null}
        selectedTo={null}
        showArrows={true}
        showToday={false}
        value={1527811200000}
        weekdays={
          Array [
            "Пн",
            "Вт",
            "Ср",
            "Чт",
            "Пт",
            "Сб",
            "Вс",
          ]
        }
      />
    </div>
  </Popup>
</span>
`;
